### DISCUSSION TITLE
How to read multi-dimensional array from previous step with js?

### DISCUSSION THREAD

**Discussion by razoons** [Budibase/budibase#8760] [Help]
Hi everyone,

I'm struggling trying to read a 2 dimensions array from a previous step.
I have a first step where I query rows from a datasource
and a second step of JS Scripting where I return a 2 dimensions array that looks like this:

>participants:
>   0:
>      'HN1': 2
>      'HN2': 5
>   1:
>      'HN1': 10
>      'HN2': 8
>   2:
>      'HN1': 9
>      'HN2': 0

Now I would like to use this array in a further step in javascript to populate a field in a new row.

But somehow, I don't know exactly how to parse this array.
With Handlebars, it would be something like:
`{{ steps.2.value.1.HN1 }}` to get the value 10.

But I would like to use javascript instead.
How could I get this value?

I tried: `return $("steps.2.value.1.HN1");` `return $("steps.2.value.[1].HN1");` `return $("steps.2.value.[1].HN1");`
I also tried by declaring the array first:
`let a=$("steps.2.value");  return a[2]['HN2'];`

but it doesn't work (I don't have any error though, it's just not processing - it uses the last value that worked)

Anyone could help me with this?
Thank you

**Comment by melohagan** [comment]
Hey @razoons 

Would you be able to share some screenshots of your automation please? An app export that reproduces the issue would also be of great help!

**Comment by razoons** [comment]
Hi @melohagan !

Indeed, will be clearer for everyone with pictures and details.

So here is my table. The goal for me is to compute the "points" based on the ranking of each "Manche" and the number of participants for this "classe".
![Capture d’écran 2022-11-22 185616](https://user-images.githubusercontent.com/44969007/203494364-23f4b28c-a95f-4b2f-94ec-312b40c8aba0.png)

I have a first step in my automation where I loop over the items to get the list
![Capture d’écran 2022-11-22 185636](https://user-images.githubusercontent.com/44969007/203494466-e50cea52-33f0-4577-84ba-c6e8a51e15ae.png)

Then I have a second step where I compute the number of participants per "manche" et per "classe"
![Capture d’écran 2022-11-22 185711](https://user-images.githubusercontent.com/44969007/203494705-e792e1fc-a009-4f67-9fbb-0acf7a68a278.png)
![Capture d’écran 2022-11-22 190124](https://user-images.githubusercontent.com/44969007/203494731-9446f267-8cf4-4f45-a3f7-b1ac93f1f677.png)

The value of this step is a 2 dimensions array (participants[manche][classe]) providing the number of participants for the "manche" and "classe".

Now I need to update each item with the points value (so using the participants[][] returned during step 2).
For now, I'm just trying to read the step2 array but this is where I struggle.
How should I read and parse the array coming from step2 in Javascript?
![Sans titre](https://user-images.githubusercontent.com/44969007/203495651-f8e7592f-cdd0-4cbd-9def-7daced9cbb47.png)
![Capture d’écran 2022-11-23 085355](https://user-images.githubusercontent.com/44969007/203495807-c115e2df-89a0-4bec-9da1-70226dd75400.png)

Thank you very much for your help!

**Reply by melohagan** [reply]
Looking at the JavaScript block, I think there are a few potential mistakes in there.

The binding should be `steps[1].rows` and not `{{ steps.1.rows }}`

I think the for loop should be `for (let i ...`

Finally you have a check for `=== 'undefined'`. Did you mean `=== undefined` ?

**Reply by razoons** [reply]
Thanks.

I made the changes for a better code  but it was not the rootcause (with my previous code, I tried to return participants[1]['HN1'] and I 
got the correct value when testing it). Still, made the changes you mentioned.

I have the feeling the step2 result is fine (not talking about the quality of the code here..) because the value is an array and if I ask to return a specific value of this array, it will give back the expected value (as per my data).

You can find attached the exported app (hope I made the correct export)
[My app-export-1669205788646.tar.gz](https://github.com/Budibase/budibase/files/10075343/My.app-export-1669205788646.tar.gz)

Here is my latest test:
Step2: asking for `return participants[1]['HN1']`
Step3: assigning `step2.value` to the attribute `points`
When testing, I can see that points takes the value `3` as expected (3 items HN1 for the "manche 1")

But as you understood, I want to return the entire array, not a specific value. But don't know how to read a specific value of `step2.value` within Step3 javascript code.

**Reply by melohagan** [reply]
I don't think the JavaScript is correct. This is what I'm seeing in the output of the JavaScript code block:

```json
{
  "success": true,
  "value": [
    null,
    [],
    [],
    [],
    [],
    [],
    []
  ]
}
```
